dash-renderer/src/actions/dependencies_ts.ts

```diff
@@ -0,0 +1,323 @@
+import {
+    all,
+    assoc,
+    concat,
+    difference,
+    filter,
+    flatten,
+    forEach,
+    isEmpty,
+    keys,
+    map,
+    mergeWith,
+    partition,
+    pickBy,
+    props,
+    reduce,
+    zipObj
+} from 'ramda';
+import { ICallback, ICallbackProperty, ICallbackDefinition, ILayoutCallbackProperty, ICallbackTemplate } from '../types/callbacks';
+import { addAllResolvedFromOutputs, splitIdAndProp, stringifyId, getCallbacksInLayout, isMultiValued, idMatch } from './dependencies';
+import { getPath } from './paths';
+
+export const DIRECT = 2;
+export const INDIRECT = 1;
+export const mergeMax = mergeWith(Math.max);
+
+export const combineIdAndProp = ({
+    id,
+    property
+}: ICallbackProperty) => `${stringifyId(id)}.${property}`;
+
+export function getCallbacksByInput(
+    graphs: any,
+    paths: any,
+    id: any,
+    prop: any,
+    changeType?: any,
+    withPriority: boolean = true
+): ICallback[] {
+    const matches: ICallback[] = [];
+    const idAndProp = combineIdAndProp({ id, property: prop });
+
+    if (typeof id === 'string') {
+        // standard id version
+        const callbacks = (graphs.inputMap[id] || {})[prop];
+        if (!callbacks) {
+            return [];
+        }
+
+        callbacks.forEach(
+            addAllResolvedFromOutputs(resolveDeps(), paths, matches)
+        );
+    } else {
+        // wildcard version
+        const keys = Object.keys(id).sort();
+        const vals = props(keys, id);
+        const keyStr = keys.join(',');
+        const patterns: any[] = (graphs.inputPatterns[keyStr] || {})[prop];
+        if (!patterns) {
+            return [];
+        }
+        patterns.forEach(pattern => {
+            if (idMatch(keys, vals, pattern.values)) {
+                pattern.callbacks.forEach(
+                    addAllResolvedFromOutputs(
+                        resolveDeps(keys, vals, pattern.values),
+                        paths,
+                        matches
+                    )
+                );
+            }
+        });
+    }
+    matches.forEach(match => {
+        match.changedPropIds[idAndProp] = changeType || DIRECT;
+        if (withPriority) {
+            match.priority = getPriority(graphs, paths, match)
+        }
+    });
+    return matches;
+}
+
+/*
+ * Take a list of callbacks and follow them all forward, ie see if any of their
+ * outputs are inputs of another callback. Any new callbacks get added to the
+ * list. All that come after another get marked as blocked by that one, whether
+ * they were in the initial list or not.
+ */
```

Comment:
obsolete comment

If I'm understanding this correctly: `priority` is an array, whose length is the number of the number of callbacks in the chain starting with this one (and including this one), that block other callbacks. The entries in that array are the number of callbacks directly blocked by each of those, but after the first one they're in a somewhat meaningless order. Is that right?

Reply:
oh no, looks like I misunderstood... the length is the depth of the whole callback tree and the entries are the number of callbacks blocked by each level of the tree?

Reply:
For a given callback with `[3, 4, 2]` it means that 3 callbacks can be triggered by the callback being evaluated, 4 callbacks can be triggered by those 3, and 2 callbacks by those 4. It's amounts to looking at the callback tree and determining how many callbacks exist at each depth in the tree.